pkg/scheduler/scheduler.go

```diff
@@ -390,79 +372,132 @@ func (s *Scheduler) scheduleNext() bool {
 	}
 	defer s.queue.Done(key)
 
-	start := time.Now()
-
-	keys := key.(string)
-
-	var err error
-	switch s.getScheduleType(keys) {
-	case FirstSchedule:
-		err = s.scheduleOne(keys)
-		klog.Infof("Start scheduling binding(%s)", key)
-		metrics.BindingSchedule(string(FirstSchedule), metrics.SinceInSeconds(start), err)
-	case ReconcileSchedule: // share same logic with first schedule
-		err = s.scheduleOne(keys)
-		klog.Infof("Reschedule binding(%s) as placement changed", keys)
-		metrics.BindingSchedule(string(ReconcileSchedule), metrics.SinceInSeconds(start), err)
-	case ScaleSchedule:
-		err = s.scaleScheduleOne(keys)
-		klog.Infof("Reschedule binding(%s) as replicas scaled down or scaled up", keys)
-		metrics.BindingSchedule(string(ScaleSchedule), metrics.SinceInSeconds(start), err)
-	case FailoverSchedule:
-		if features.FeatureGate.Enabled(features.Failover) {
-			err = s.rescheduleOne(keys)
-			klog.Infof("Reschedule binding(%s) as cluster failure", keys)
-			metrics.BindingSchedule(string(FailoverSchedule), metrics.SinceInSeconds(start), err)
-		}
-	case AvoidSchedule:
-		klog.Infof("Don't need to schedule binding(%s)", keys)
-	default:
-		err = fmt.Errorf("unknow schedule type")
-		klog.Warningf("Failed to identify scheduler type for binding(%s)", keys)
-	}
-
+	err := s.doSchedule(key.(string))
 	s.handleErr(err, key)
 	return true
 }
 
-func (s *Scheduler) scheduleOne(key string) (err error) {
-	klog.V(4).Infof("begin scheduling ResourceBinding %s", key)
-	defer klog.V(4).Infof("end scheduling ResourceBinding %s: %v", key, err)
-
+func (s *Scheduler) doSchedule(key string) error {
 	ns, name, err := cache.SplitMetaNamespaceKey(key)
 	if err != nil {
 		return err
 	}
+	if len(ns) > 0 {
+		return s.doScheduleBinding(ns, name)
+	}
+	return s.doScheduleClusterBinding(name)
+}
 
-	if ns == "" {
-		clusterResourceBinding, err := s.clusterBindingLister.Get(name)
+func (s *Scheduler) doScheduleBinding(namespace, name string) error {
+	rb, err := s.bindingLister.ResourceBindings(namespace).Get(name)
+	if err != nil {
 		if apierrors.IsNotFound(err) {
+			// the binding does not exist, do nothing
 			return nil
 		}
+		return err
+	}
 
-		clusterPolicyName := util.GetLabelValue(clusterResourceBinding.Labels, policyv1alpha1.ClusterPropagationPolicyLabel)
+	start := time.Now()
+	if !helper.IsBindingReady(&rb.Status) {
```

Comment:
hi, the `Scheduled` condition is just to indicate that the ResourceBinding or ClusterResourceBinding has been scheduled once, not to indicate that the status of every scheduling process, right?

Reply:
Yes, please refer to https://github.com/karmada-io/karmada/issues/821#issuecomment-944842382

Reply:
> 

It is out of my expectation, I used to think that this condition is to indicate the status of every scheduling process. If first schedule, rescedule or scale schedule failed, it should be set to `False`. For now, we will keep the schedule result of last success scheduling if reschedule failed. But for users, they don't know why schedule failed.

Reply:
> 

/cc @RainbowMango Is it the expected  behavior of this condition?

Reply:
We do need a condition to reflect the schedule failed state.  I thought it might be too complicated, so we didn't handle the failed case in https://github.com/karmada-io/karmada/issues/821.

Reply:
> 

If so, may be we can't used this condition to check if it is the `FirstSchedule`, because it may be set to false when schedule failed. @dddddai

Reply:
Do we need to care if it is "First" schedule? If not,I think we may drop `FirstSchedule` and use another name instead(e.g. `Schedule`).

Reply:
> 

uh-huh, it seems like the "First" schedule is a kind of special "Reschedule"

Reply:
Oh there is a problem: for example, if `FailoverSchedule` failed, then `Scheduled` condition would be false, then scheduler would regard it as `FirstSchedule` whose behavior is **different** from `FailoverSchedule`

Reply:
> 

How about to do it like:
```
func (s *Scheduler) doScheduleBinding(namespace, name string) error {
	rb, err := s.bindingLister.ResourceBindings(namespace).Get(name)
	if err != nil {
		if apierrors.IsNotFound(err) {
			// the binding does not exist, do nothing
			return nil
		}
		return err
	}

	start := time.Now()

	defer func() {
		if err != nil {
			// set "Scheduled" condition false
		} else {
			// set "Scheduled" condition true
		}
	}()

	policyPlacement, policyPlacementStr, err := s.getPlacement(rb)
	if err != nil {
		return err
	}
	appliedPlacement := util.GetLabelValue(rb.Annotations, util.PolicyPlacementAnnotation)
	if policyPlacementStr != appliedPlacement {
		// policy placement changed, need reschedule
		klog.Infof("Reschedule ResourceBinding(%s/%s) as placement changed", namespace, name)
		err = s.scheduleResourceBinding(rb)
		metrics.BindingSchedule(string(ReconcileSchedule), metrics.SinceInSeconds(start), err)
		return err
	}
```
`Scheduled` condition is to indicate the schedule result of every schedule process. If schedule failed, we won't update the  applied-placement annation, but will set the `Scheduled` condition as false. The scheduler behavior will be consistent. The schedule will `scheduleResourceBinding` once applied policy is not the same as current policy. Otherwise, perform failover schedule or scale schedule.

Reply:
Yeah, it sounds reasonable, i think we can get this one merged then do this job in #988

Reply:
> 

Yeah, I agree